topics/atomic-slot-migration.md

```diff
@@ -0,0 +1,258 @@
+This command can be sent to the whole cluster to cancel all slot migrations
+everywhere.
+
+For more details on `CLUSTER CANCELSLOTMIGRATIONS` see the
+[command documentation](../commands/cluster-cancelslotmigrations.md).
+
+## Behind the scenes of Atomic Slot Migration
+
+Atomic Slot Migration utilizes a completely different process than
+`CLUSTER SETSLOT`-based migrations:
+
+1. Immediately after `CLUSTER MIGRATESLOTS` is received by the source node, it
+   initiates a connection to the target node and performs authentication,
+   similar to how a replication link is initialized.
+2. Once established, the source node uses a new internal command -
+   `CLUSTER SYNCSLOTS` - to inform the target of the migration.
+3. The source node then forks a child process to do a one-time snapshot of the
+   slot contents. The fork iterates all hash slots and serializes their contents
+   over the slot migration link. The contents are subsequently replicated to any
+   replicas of the target node.
+4. While the child process is doing the snapshot, the parent process tracks all
+   mutations performed on the migrating hash slots.
+5. Once the child process snapshot finishes, the parent process sends all
+   accumulated mutations. Any new mutations received during this step are also
+   sent.
+6. Once the amount of in-flight mutations goes below a configured threshold, the
+   parent process freezes mutations temporarily to allow final synchronization
+   of the hash slots.
+7. Once the target node is completely caught up, it takes over the hash slots
+   and broadcasts ownership to the cluster
+8. When the source node finds out about the migration, it deletes the keys in
+   the hash slot and unfreezes mutations. Clients will now get `MOVED`
+   redirections to the target node, which now owns the hash slots. The slot
+   migration is completed.
+
+### Isolation of importing hash slots from clients
+
+Since slot ownership is not moved until the very end of the migration, commands
+targeting migrating hash slots on the target node will receive `MOVED`
+redirections per the cluster specification. But there are some commands that
+operate on the entire database:
+
+1. `KEYS`/`SCAN`: These commands allow a client to list out all keys on a shard.
+2. `DBSIZE`/`INFO`: These commands provide statistical information about how
+   many keys are on a shard.
+3. `FLUSHDB`/`FLUSHALL`: These commands allow a client to drop all data in a
+   database, or on all databases, on a node.
+
+To handle this, all importing hash slots are marked specially and hidden from
+read operations on both the target primary and the target replica.
+
+`FLUSHDB` and `FLUSHALL` present a special case where we fail the slot migration
+when being executed on **both the source and target node**. It is expected that
+operators would retry the migration after flushing, which should now succeed
+almost instantly due to an empty database.
+
+### Deep Dive on `CLUSTER SYNCSLOTS`
```

Comment:
Might be my bias, but I don't like so much detail in documentation. It's likely to get stale, and ultimately provides **too** much information to end users. I think this is more interesting for a blog, which will be point in time.